dnf/base.py

```diff
@@ -881,6 +884,8 @@ def display_banner(pkg, count):
             self.history.end(rpmdbv, 0)
         timer()
         self._trans_success = True
+        if ret:
+            raise dnf.exceptions.Error('Transaction verification fails')
```

Comment:
This whole block is not needed since you can extract failed transaction info directly from transaction at the bottom of `BaseCli.do_transaction()`. 

I still have a feeling that this should be better handled on rpm level somehow. Please ask Florian on his opinion in the attached bugzilla.

And my last concert is the content of dnf database. We definitely shouldn't mark failed packages as installed.